YololCompetition/Services/Messages/DBMessages.cs

```diff
@@ -0,0 +1,106 @@
+        private readonly BaseSocketClient _client;
+
+        public enum messageType : uint {Current, Leaderboard}
+        
+        public DbMessages(IDatabase database, ICron cron, IChallenges challenges, BaseSocketClient client)
+        {
+            _database = database;
+            _cron = cron;
+            _challenges = challenges;
+            _client = client;
+
+            try
+            {
+                _database.Exec("CREATE TABLE IF NOT EXISTS `Messages` (`ChannelID` INTEGER NOT NULL, `MessageID` INTEGER NOT NULL, `ChallengeID` INTEGER NOT NULL, 'MessageType' INTEGER NOT NULL, PRIMARY KEY(`MessageID`, `ChallengeID'))");
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
+            }
+        }
+
+        public async Task TrackMessage(ulong ChannelID, ulong MessageID, ulong ChallengeID, uint MessageType)
+        {
+            await using var cmd = _database.CreateCommand();
+            cmd.CommandText = "INSERT into Messages (ChannelID, MessageID, ChallengeID, MessageType) values(@ChannelID, @MessageID, @ChallengeID, @MessageType)";
+            cmd.Parameters.Add(new SqliteParameter("@ChannelID", DbType.UInt64) { Value = ChannelID });
+            cmd.Parameters.Add(new SqliteParameter("@MessageID", DbType.UInt64) { Value = MessageID });
+            cmd.Parameters.Add(new SqliteParameter("@ChallengeID", DbType.UInt64) { Value = ChallengeID });
+            cmd.Parameters.Add(new SqliteParameter("@MessageType", DbType.UInt64) { Value = MessageType });
+            await cmd.ExecuteNonQueryAsync();
+        }
+
+        public IAsyncEnumerable<Message> GetCurrentMessages(ulong challengeID)
+        {
+            DbCommand PrepareQuery(IDatabase database)
+            {
+                var cmd = _database.CreateCommand();
+                cmd.CommandText = "SELECT from 'Messages' WHERE 'ChallengID' = @challengeID";
+                cmd.Parameters.Add(new SqliteParameter("@ChallengeID", DbType.UInt64) { Value = challengeID});
+                return cmd;
+            }
+
+            return new SqlAsyncResult<Message>(_database, PrepareQuery, ParseMessage);
+        }
+
+        private static Message ParseMessage(DbDataReader reader)
+        {
+            return new Message(
+                ulong.Parse(reader["ChannelID"].ToString()!),
+                ulong.Parse(reader["MessageID"].ToString()!),
+                ulong.Parse(reader["ChallengeID"].ToString()!),
+                uint.Parse(reader["MessageType"].ToString()!)
+            );
+        }
+
+        public void StartMessageWatch()
+        {
+            _cron.Schedule(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(1), uint.MaxValue, async () => {
+                // Get current challenge
+                var c = await _challenges.GetCurrentChallenge();
+
+                var messages = GetCurrentMessages(c.Id);
+                await foreach (Message entry in messages) 
```

Comment:
Maybe wrap every case in a `try/catch` block and log errors - to ensure the cron job never dies.

Reply:
I need help with this one. I don't know what I'm doing with error handling and there's a few places I would like to implement it.

Reply:
At the moment Referee does not have a structured logging system in place, so if you want to log out an error just call `Console.WriteLine(blah);`.

If you're unfamiliar with `try/catch` you can wrap a block of code:

```
try
{
    SomeCodeThatMightThrow();
}
catch (Exception e)
{
    HandleError(e);
}
```

This stops the exception propagating upwards because it stops in the `catch` block.

So in this case, I'm just suggesting that you put everything inside the foreach loop in a `try/catch` in a block, that way if any case fails it won't terminate the loop and the entire cron job. You _could_ do slightly more advanced error handling than just logging it e.g. remove that particular entry from the DB or wait a short while and retry this case. Since all entries are wiped from this table every 5 days I think it's good enough just to log and move on.